File: cronosFacial/app/src/main/java/com/cronosedx/cronosfacial/analyzer/GazeEstimator.kt
package com.cronosedx.cronosfacial.analyzer

import android.util.Log
import kotlin.math.abs

/**
 * Estimates gaze direction based on facial landmarks.
 * 
 * Uses eye position and iris location (simulated from landmarks) to determine
 * where the user is looking.
 */
class GazeEstimator {
    private val TAG = "GazeEstimator"
    
    // Threshold values for gaze detection
    private val GAZE_THRESHOLD_X = 0.15f
    private val GAZE_THRESHOLD_Y = 0.15f
    
    /**
     * Estimates gaze direction from facial landmarks.
     * 
     * @param landmarks Array of facial landmarks (468 points for MediaPipe Face Mesh)
     * @return Gaze direction: "center", "left", "right", "up", "down", "up-left", etc.
     */
    fun estimateGaze(landmarks: FloatArray): String {
        if (landmarks.isEmpty()) {
            Log.w(TAG, "No landmarks provided")
            return "unknown"
        }
        
        Log.d(TAG, "Estimating gaze from ${landmarks.size} landmarks")
        
        // Simulate eye center and iris positions from landmarks
        // In a real implementation, we would extract specific eye landmarks
        // For MediaPipe: Left eye indices: 33, 133, 160, 159, 158, 157, 173, 144
        //                Right eye indices: 362, 263, 387, 386, 385, 384, 398, 373
        
        // Simulate horizontal and vertical gaze offset
        // In real implementation, calculate iris position relative to eye center
        val gazeOffsetX = calculateHorizontalGaze(landmarks)
        val gazeOffsetY = calculateVerticalGaze(landmarks)
        
        Log.d(TAG, "Gaze offset - X: $gazeOffsetX, Y: $gazeOffsetY")
        
        // Determine gaze direction based on offsets
        return determineGazeDirection(gazeOffsetX, gazeOffsetY)
    }
    
    /**
     * Calculate horizontal gaze direction.
     * Positive = looking right, Negative = looking left
     */
    private fun calculateHorizontalGaze(landmarks: FloatArray): Float {
        // Simulate gaze calculation using landmark variation
        // In practice, this would use actual eye and iris landmark positions
        val variation = (landmarks.sum() % 1.0f) - 0.5f
        return variation * 0.6f // Scale to reasonable range
    }
    
    /**
     * Calculate vertical gaze direction.
     * Positive = looking up, Negative = looking down
     */
    private fun calculateVerticalGaze(landmarks: FloatArray): Float {
        // Simulate vertical gaze using landmark data
        val variation = ((landmarks.sum() * 1.3f) % 1.0f) - 0.5f
        return variation * 0.5f // Scale to reasonable range
    }
    
    /**
     * Determine gaze direction string from X/Y offsets.
     */
    private fun determineGazeDirection(offsetX: Float, offsetY: Float): String {
        val horizontal = when {
            offsetX > GAZE_THRESHOLD_X -> "right"
            offsetX < -GAZE_THRESHOLD_X -> "left"
            else -> null
        }
        
        val vertical = when {
            offsetY > GAZE_THRESHOLD_Y -> "up"
            offsetY < -GAZE_THRESHOLD_Y -> "down"
            else -> null
        }
        
        return when {
            horizontal != null && vertical != null -> "$vertical-$horizontal"
            horizontal != null -> horizontal
            vertical != null -> vertical
            else -> "center"
        }
    }
    
    /**
     * Get confidence score for gaze estimation (0-1).
     */
    fun getGazeConfidence(landmarks: FloatArray): Float {
        if (landmarks.isEmpty()) return 0f
        
        // In real implementation, confidence would be based on:
        // - Face detection confidence
        // - Eye openness
        // - Landmark quality
        return 0.85f
    }
}

File: cronosFacialIOS/CronosFacial/Analyzers/GazeEstimator.swift
//
//  GazeEstimator.swift
//  CronosFacial
//
//  Gaze direction estimation using eye landmarks
//

import Foundation
import Vision

/// Estimates gaze direction from facial landmarks
class GazeEstimator {
    // Threshold values for gaze detection
    private let gazeThresholdX: Float = 0.15
    private let gazeThresholdY: Float = 0.15
    
    /// Estimates gaze direction from landmark array
    /// - Parameter landmarks: Array of facial landmarks
    /// - Returns: Gaze direction string
    func estimateGaze(from landmarks: [Float]) -> String {
        guard !landmarks.isEmpty else {
            print("[GazeEstimator] No landmarks provided")
            return "unknown"
        }
        
        print("[GazeEstimator] Estimating gaze from \(landmarks.count) landmarks")
        
        let gazeOffsetX = calculateHorizontalGaze(from: landmarks)
        let gazeOffsetY = calculateVerticalGaze(from: landmarks)
        
        print("[GazeEstimator] Gaze offset - X: \(gazeOffsetX), Y: \(gazeOffsetY)")
        
        return determineGazeDirection(offsetX: gazeOffsetX, offsetY: gazeOffsetY)
    }
    
    /// Estimates gaze from VNFaceObservation
    /// - Parameter faceObservation: Face observation from Vision
    /// - Returns: Gaze direction string
    func estimateGaze(from faceObservation: VNFaceObservation) -> String {
        guard let landmarks = faceObservation.landmarks else {
            return "center"
        }
        
        // Extract eye positions
        guard let leftPupil = landmarks.leftPupil,
              let rightPupil = landmarks.rightPupil else {
            return "center"
        }
        
        // Calculate gaze based on pupil position relative to eye center
        let leftPupilPoints = leftPupil.normalizedPoints
        let rightPupilPoints = rightPupil.normalizedPoints
        
        guard !leftPupilPoints.isEmpty, !rightPupilPoints.isEmpty else {
            return "center"
        }
        
        // Average pupil positions
        let avgPupilX = (leftPupilPoints[0].x + rightPupilPoints[0].x) / 2
        let avgPupilY = (leftPupilPoints[0].y + rightPupilPoints[0].y) / 2
        
        // Determine gaze based on pupil position
        // Values are normalized, so 0.5 is center
        let offsetX = Float(avgPupilX - 0.5)
        let offsetY = Float(avgPupilY - 0.5)
        
        return determineGazeDirection(offsetX: offsetX * 2, offsetY: offsetY * 2)
    }
    
    /// Calculate horizontal gaze direction
    private func calculateHorizontalGaze(from landmarks: [Float]) -> Float {
        // Simulate gaze calculation using landmark variation
        let variation = (landmarks.reduce(0, +).truncatingRemainder(dividingBy: 1.0)) - 0.5
        return variation * 0.6 // Scale to reasonable range
    }
    
    /// Calculate vertical gaze direction
    private func calculateVerticalGaze(from landmarks: [Float]) -> Float {
        // Simulate vertical gaze using landmark data
        let variation = ((landmarks.reduce(0, +) * 1.3).truncatingRemainder(dividingBy: 1.0)) - 0.5
        return variation * 0.5 // Scale to reasonable range
    }
    
    /// Determine gaze direction string from X/Y offsets
    private func determineGazeDirection(offsetX: Float, offsetY: Float) -> String {
        let horizontal: String? = {
            if offsetX > gazeThresholdX {
                return "right"
            } else if offsetX < -gazeThresholdX {
                return "left"
            }
            return nil
        }()
        
        let vertical: String? = {
            if offsetY > gazeThresholdY {
                return "up"
            } else if offsetY < -gazeThresholdY {
                return "down"
            }
            return nil
        }()
        
        // Combine directions
        if let h = horizontal, let v = vertical {
            return "\(v)-\(h)"
        } else if let h = horizontal {
            return h
        } else if let v = vertical {
            return v
        } else {
            return "center"
        }
    }
    
    /// Get confidence score for gaze estimation
    /// - Parameter landmarks: Array of facial landmarks
    /// - Returns: Confidence score (0-1)
    func getGazeConfidence(from landmarks: [Float]) -> Float {
        guard !landmarks.isEmpty else {
            return 0.0
        }
        
        // In real implementation, confidence would be based on:
        // - Face detection confidence
        // - Eye openness
        // - Landmark quality
        return 0.85
    }
}
